_posts/2021-01-08-consumer-tuning.md

```diff
@@ -0,0 +1,292 @@
+
+Throughput is improved by increasing the amount of data fetched in a batch, but with some cost to latency. Consequently, adjusting these properties lower has the effect of lowering end-to-end latency. We'll look a bit more at targeting latency by increasing batch sizes in the next section.
+
+You can use one or both of these properties. If you use both, Kafka will respond to a fetch request when the first of either threshold is reached.
+
+```properties
+# ...
+fetch.max.wait.ms=500
+fetch.min.bytes=16384
+# ...
+```
+
+## Lowering latency by increasing the maximum batch sizes
+
+Increasing the minimum amount of data fetched in a request can help with increasing throughput. But if you want to do something to improve latency, you can extend your thresholds by increasing the _maximum_ amount of data that can be fetched by the consumer from the broker.
+
+* **`fetch.max.bytes`**
+Sets a maximum limit in bytes on the amount of data fetched from the broker at one time
+
+* **`max.partition.fetch.bytes`**
+Sets a maximum limit in bytes on how much data is returned for each partition, which must always be larger than the number of bytes set in the broker or topic configuration for `max.message.bytes`.
+
+By increasing the values of these two properties, and allowing more data in each request, latency might be improved as there are fewer fetch requests.
+
+```properties
+# ...
+fetch.max.bytes=52428800
+max.partition.fetch.bytes=1048576
+# ...
+```
+
+> Be careful with memory usage. The maximum amount of memory a client can consume is calculated approximately as:
+>
+> <pre><i>NUMBER-OF-BROKERS</i> * fetch.max.bytes and <i>NUMBER-OF-PARTITIONS</i> * max.partition.fetch.bytes</pre>
+
+
+## Avoiding data loss or duplication when committing offsets
+
+For applications that require durable message delivery, you can increase the level of control over consumers when committing offsets to minimize the risk of data being lost or duplicated.
+
+Kafka's _auto-commit mechanism_ is pretty convenient (and sometimes suitable). When enabled, consumers commit the offsets of messages automatically. But convenience, as always, has a price. By allowing your consumer to commit offsets, you are introducing a risk of data loss and duplication. Why? Quite easily as it happens. You see, the auto-commit can _only_ avoid data loss if all messages are processed before the next poll of the broker or the consumer closes. Using the auto-commit, the consumer will commit offsets received from the broker at 5000ms intervals. In that time, the inherent constraints of auto-committing might lead to:
```

Comment:
@tombentley you are changing with this:

> When enabled, consumers commit the offsets of the just-processed messages automatically when `poll()` is called to get the next messages to process

it sounds that if my application has auto commit every 3 seconds but it does a poll every 30 seconds, the offsets will not be committed after 3 seconds but it will wait 30 seconds (the next poll). Aren't these two things not related? I mean, isn't auto commit async compared to the poll? Otherwise the `auto.commit.interval.ms` doesn't make so much sense, or?

Reply:
@ppatierno ah, you're quite right! I do apologise. Is this better:

```suggestion
Kafka's _auto-commit mechanism_ is pretty convenient (and sometimes suitable, depending on the use case). When enabled, consumers commit the offsets of messages automatically every `auto.commit.interval.ms` milliseconds. But convenience, as always, has a price. By allowing your consumer to commit offsets, you are introducing a risk of data loss and duplication. How? 

* If your application commits an offset and then crashes before all the messages up to that offset have actually been processed those messages won't get processed when the application is restarted (data loss).
* If your application has processed all the messages and then crashes before the offset is committed automatically then the last offset will be used when the application restarts and you'll process those messages again. 

Turning off the auto commit functionality can help with the first issue, because you can write your code on only commit offsets once messages have actually been processed. But manual commit can't completely eliminate the second problem because you can't guarantee that the offset commit message will always be processed by the broker. This is the case even if you do a synchronous offset commit after every processing each message.
```

?

Reply:
I updated the first bit as follows:

![image](https://user-images.githubusercontent.com/47596553/103876626-c8a87b00-50cb-11eb-99df-308a1e84e36f.png)

I added the final paragraph to sum up the section:

![image](https://user-images.githubusercontent.com/47596553/103876817-0ad1bc80-50cc-11eb-80bd-e3794f58c50f.png)